File: python/prime-factors/prime_factors.py
from math import sqrt
def prime_factors(natural_number):
    if natural_number < 2:
        return []
    for i in range(2, int(sqrt(natural_number)) + 1):
        if natural_number % i == 0:
            return [i] + prime_factors(natural_number // i)
    return [natural_number]

File: haskell/prime-factors/src/PrimeFactors.hs
module PrimeFactors (primeFactors) where

primeFactors :: Integer -> [Integer]
primeFactors 0 = []
primeFactors 1 = []
primeFactors n = 
  let factors = filter (\i -> n `mod` i == 0) [2..(round . sqrt . fromInteger) n]
  in case factors of
    [] -> [n]
    (x:xs) -> x : (primeFactors (n `div` x))
